pipelines/rbsa/tasks/group_sites.py

```diff
@@ -0,0 +1,136 @@
+
+            if zipcode3digit in zipdf_dict.keys(): 
+                zipdf_dict[zipcode3digit] = self.add_df(zipdf_dict[zipcode3digit], site_df)
+            else:
+                zipdf_dict[zipcode3digit] = site_df
+
+        # make single dataframe out op dictionary
+        initialization = True
+
+        for zip3 in zipdf_dict.keys():
+            zip_df = zipdf_dict[zip3]
+            zip_df.insert(loc=0, column='zipcode', value=zip3)
+
+            if initialization:
+                area_loads = zip_df
+            else:
+                area_loads.append(zip_df,ignore_index=False, sort=False)
+
+        self.validate(area_loads)
+
+        area_loads.to_csv('local_data/area_loads.csv')
+
+    
+    def get_zipsitemapping(self, allsites, site_zipmap):
+        """creates new dict that maps 3 digit zipcodes to sites in zipcode
+        Currenly unused
+        """
+
+        zip_sitemap = {}
+
+        for site in allsites:
+            zipcode = site_zipmap[site]
+            zipcode3digit = zipcode[:3]
+            
+            if zipcode3digit in zip_sitemap.keys(): 
+                zip_sitemap[zipcode3digit].append(site)
+            else:
+                zip_sitemap[zipcode3digit] = [site]
+
+        return zip_sitemap
+
+    def get_zip5_dict(self, allsites, site_zipmap):
+        """maps 3 digit zipcodes to list of 5 digit zipcodes, for correlation
+        """   
+
+        zipcode5map = {}
+
+        for site in allsites:
+            zipcode = site_zipmap[site]
+            zipcode3digit = zipcode[:3]
+            
+            if zipcode3digit in zipcode5map.keys(): 
+                zipcode5map[zipcode3digit].append(zipcode)
+            else:
+                zipcode5map[zipcode3digit] = [zipcode]
+
+        return zipcode5map
+
+    def add_df(self, df1, df2):
+        """This function will add cell values of two dataframes.
+        """
+
+        if (df1.index.min() < df2.index.min()) & (df1.index.max() >= df2.index.max()):
+            df2= df2.reindex_like(df1).fillna(0)
+        elif (df1.index.min() <= df2.index.min()) & (df1.index.max() > df2.index.max()):
+            df2= df2.reindex_like(df1).fillna(0)
+
+        elif (df1.index.min() > df2.index.min()) & (df1.index.max() <= df2.index.max()):
+            df1= df1.reindex_like(df2).fillna(0)
+        elif (df1.index.min() >= df2.index.min()) & (df1.index.max() < df2.index.max()):
+            df1= df1.reindex_like(df2).fillna(0)
+
+        elif (df1.index.min() != df2.index.min()) & (df1.index.max() != df2.index.max()):
+            df1= df1.reindex_like(df2).fillna(0)
+            df2= df2.reindex_like(df1).fillna(0)
+
+        df_add = df1.fillna(0) + df2.fillna(0)
+
+        return df_add
+
+    def validate(self, df):
+        """ Validation
+        """
+        if df.isnull().values.any():
+            err_msg = ('Error found during grouping of sites to zip codes.')
+            logger.exception(err_msg)
+            raise ValueError(err_msg)
```

Comment:
No need to change anything here, but I will change how validations fail. Instead of raising an exception, let's set a flag on the task class, something like `is_successful`, so the parent pipeline can inspect and choose to continue execution of the pipeline or completely stop and perform cleanup. Again, nothing to change here, but mostly a comment for me.